### ISSUE TITLE
Different metrics when loading weights from saved checkpoints between `main` and PR #166 despite identical final in-memory models

### ISSUE THREAD

**Issue by jeipollack** [CosmoStat/wf-psf#168]
When loading model weights from saved checkpoints, the evaluation metrics differ between `main` and #166 run with the same dataset and configuration, even though the final models in memory (after training completes) are identical.

The difference occurs only when loading weights from saved checkpoints, not when evaluating directly from the final in-memory trained model.

Using a weight comparison function:

```
def diff_models(m1, m2, atol=1e-8):
    for w1, w2 in zip(m1.weights, m2.weights):
        if not tf.reduce_all(tf.math.abs(w1 - w2) <= atol):
            print(f"↯  {w1.name} differs")
            return
    print("✅  all weights identical")
```
shows weight differences between models loaded from checkpoints of the two branches (version: 0 differs), despite final in-memory weights being identical.

It seems this might be due to changing how the model is compiled during training even though the same parameters were passed.  In new branch, a helper function was added and this could lead to changes in different variable names and structure for the optimizer state

`main`
```
# Define Loss
    if param_loss is None:
        loss = tf.keras.losses.MeanSquaredError()
    else:
        loss = param_loss

    # Define optimisers
    if param_optim is None:
        optimizer = tf.keras.optimizers.Adam(
            learning_rate=learning_rate_param,
            beta_1=0.9,
            beta_2=0.999,
            epsilon=1e-07,
            amsgrad=False,
        )
    else:
        optimizer = param_optim

    # Define metrics
    if param_metrics is None:
        metrics = [tf.keras.metrics.MeanSquaredError()]
    else:
        metrics = param_metrics
```

in the refactored branch, a helper function was added

```
optimizer, loss, metrics = configure_optimizer_and_loss(...)
```

**Fix:**  To ensure backwards compatibility make optimizer creation consistent (i.e. restore to direct creation)

**Comment by jeipollack** [comment]
I made these changes to `train_utils.py` which should be the equivalent of what was done in `main` but short-circuited to reduce the number of lines.

```
     # Parametric  part
-    optimizer, loss, metrics = configure_optimizer_and_loss(
-        learning_rate_param, param_optim, param_loss, param_metrics
+    loss = param_loss or tf.keras.losses.MeanSquaredError()
+
+    optimizer = param_optim or tf.keras.optimizers.Adam(
+        learning_rate=learning_rate_param,
+        beta_1=0.9,
+        beta_2=0.999,
+        epsilon=1e-07,
+        amsgrad=False,
     )
+
+    metrics = param_metrics or [tf.keras.metrics.MeanSquaredError()]
```
and likewise for the non-parametric part
```
     # Non-parametric part
-    optimizer, loss, metrics = configure_optimizer_and_loss(
-        learning_rate_non_param,
-        non_param_optim,
-        non_param_loss,
-        non_param_metrics,
-        is_parametric=False,
+    loss = non_param_loss or tf.keras.losses.MeanSquaredError()
+
+    optimizer = non_param_optim or tf.keras.optimizers.Adam(
+        learning_rate=learning_rate_nonparam,
+        beta_1=0.9,
+        beta_2=0.999,
+        epsilon=1e-07,
+        amsgrad=False,
     )

+    metrics = non_param_metrics or [tf.keras.metrics.MeanSquaredError()]
```

I then relaunched v0xx and v1xx.

**Comment by jeipollack** [comment]
The metrics for `v0_1_0` appear unchanged and the checkpoints still differ. However when I compared the checkpoint weights between `main` and `v0_2_0` (unitary masks) they are identical both with helper method and the `short-circuit` direct creation of the optimizer, loss and metrics.  Here, though, the `monitor = "loss"` and not "mean_squared_error" as in `main` and `v0_1_0` runs with the #166 branch.

The only change I can spot is how masking is done, so I checked out an earlier commit state before `masked_mse` was introduced and launched wavediff training and metrics evaluation.

**Comment by jeipollack** [comment]
Update on checkpoint mismatch (vanilla training)

After comparing recent runs: `main`, `gh-165 v0_1_0`, `gh-165 v0_2_0` and `case-study early commit state before mask_mse`, I observed the following 

* `mask_mse` runs (with a unitary mask) which monitors the `loss` produces matching checkpoints with `main`.
* Vanilla runs (no mask) diverge in cycle 2, with `mean_squared_error` plateauing at epoch 2 instead of epoch 116 as in `main` and earlier commit.
`model.compile()` appears to succeed, but `model.metrics_names` is unexpectedly empty.   This might break checkpoint monitoring logic.
The issue likely stems from changes introduced after mask_mse support (e.g., conditional logic affecting metric registration).

I am going to try `git bisect` post-mask_mse to see if I can find when the behaviour changed.

**Comment by jeipollack** [comment]
I deployed the following runs using the paper's 2000 training dataset and configuration settings with the `old_main_random_seed` branches serving as a baseline:

- main : tagged 2.0.1
  - ✅ Metrics for test and train match `old_main_random_seed` 
- commit 52c9d79 : before `train_utils.py` refactor and before the refactoring to include `masked_mse` loss function 
  - ✅ Metrics for test and train match `old_main_random_seed`  
- commit 495c3da : after `train_utils.py` refactor and before the refactoring to include `masked_mse` loss function
  - ✅ Metrics for test and train match `old_main_random_seed`  
- commit 3566cd4 : after refactoring to include `masked_mse` loss function
  - ❌ Stuck in local minimum (epoch 1) in the second cycle of non-parametric training
- case_study latest : local refactor moving duplicated calls to set loss, metrics, monitor, and output outside of the multi-cycle while loop in `train.py`
   - ❌ Stuck in local minimum (epoch 1) in the second cycle of parametric training

**Comment by jeipollack** [comment]
It seems the problem came from the enhancement that introduced the `masked_mse` loss function. Specifically, the same `metrics` object was reused across both the parametric and non-parametric training phases during each cycle of the multi-cycle training loop. Previously, the loss and metrics were reset to `None` per cycle, which avoided this issue.

Apparently, in TensorFlow, objects like `tf.keras.metrics.MeanSquaredError()` are stateful: they internally track `sum_squared_errors` and `total_examples` across batches. If reused across `model.fit()` calls without resetting or re-instantiating, these accumulated states persist across training cycles.

As a result, metrics such as `mean_squared_error` may reflect stale or misleading values. Worse, if `monitor="mean_squared_error"` is used in callbacks like `ModelCheckpoint`, this can trigger premature convergence or checkpointing based on incorrect metrics.

A [conditional block](https://github.com/CosmoStat/wf-psf/blob/bae2837cca817d2482ead1341b0263117e3e9c12/src/wf_psf/training/train.py#L361-L381) inside the multi-cycle training loop was introduced to switch between `masked_mse` and standard MSE losses. This block defined loss, metrics, and outputs once per run, outside the scope of each `model.fit()` call, meaning:
```
metrics = [tf.keras.metrics.MeanSquaredError()]  # or Masked variant
```
This list of metric objects was reused across multiple calls to `model.fit()` for both parametric and non-parametric updates — causing state leakage.


**Fix**
Move the loss/metric/output initialization into a factory function that is called inside the training cycle loop. This ensures a fresh instance of loss and metric objects per cycle, per phase:

```
# inside the while loop, per cycle
loss, param_metrics, non_param_metrics, monitor, outputs, output_val = get_loss_metrics_monitor_and_outputs(
    training_handler, data_conf
)
```

**Comment by jeipollack** [comment]
This fix now reproduces the results from `main_random_seed`.
